File: dotnet/azure/Mcma.Azure.Client/AzureAd/AzureManagedIdentityAccessTokenProvider.cs
using System;
using System.Threading.Tasks;
using Mcma.Client.AccessTokens;
using Microsoft.Azure.Services.AppAuthentication;

namespace Mcma.Azure.Client.AzureAd
{
    public class AzureManagedIdentityAccessTokenProvider : IAccessTokenProvider<AzureAdAuthContext>
    {
        private AzureServiceTokenProvider AzureServiceTokenProvider { get; } = new AzureServiceTokenProvider();

        public async Task<AccessToken> GetAccessTokenAsync(AzureAdAuthContext authContext)
        {
            authContext.ValidateScope();

            var scopeAsUrl = new Uri(authContext.Scope);

            var resource = $"{scopeAsUrl.Scheme}://{scopeAsUrl.Host}";

            var authResult = await AzureServiceTokenProvider.GetAuthenticationResultAsync(resource);

            return new AccessToken
            {
                Token = authResult.AccessToken,
                ExpiresOn = authResult.ExpiresOn
            };
        }
    }
}

File: azure/client/lib/azure-ad/managed-identity/azure-managed-identity-access-token-provider.ts
import { AccessTokenProvider, AccessToken } from "@mcma/client";
import { ManagedIdentityCredential } from "@azure/identity";
import { AzureAdAuthContext } from "../azure-ad-auth-context";

export class AzureManagedIdentityAccessTokenProvider implements AccessTokenProvider<AzureAdAuthContext> {
    private identityCredential = new ManagedIdentityCredential();
    
    async getAccessToken(authContext: AzureAdAuthContext): Promise<AccessToken> {
        const token = await this.identityCredential.getToken(authContext.scope);
        return {
            accessToken: token.token,
            expiresOn: token.expiresOnTimestamp
        };
    }
}